File: Data_Structures/LinkedList/Reverse_List/reverse_linked_list.java
/* Java code to reverse a linked list

author: AnkitaM84
*/

package hacktober;

import java.util.Scanner;

class node
{
    public int data;
    node next;
    node(int d)
    {
        data = d;
        next = null;
    }
}

public class reverse_linked_list {

	public static node takeInput(Scanner s)
	{
	    System.out.println("Enter the data for nodes: ");
	    int data = s.nextInt();
	    node head = null;
	    while(data != -1)			// linked list keeps on getting created till user enters -1
	    {
	        node newNode = new node(data);
	        if(head == null)
	        {
	            head = newNode;
	        }
	        else
	        {
	            node temp = head;
	            while(temp.next != null)
	            {
	                temp = temp.next;
	            }
	            temp.next = newNode;
	        }
	        data = s.nextInt();
	    }
	    return head;
	}
	
	public static void printLL(node head)
	{
	    node temp = head;
	    while(temp != null)
	    {
	        System.out.print(temp.data + " ");
	        temp = temp.next;
	    }
	}

	public static void main(String[] args)
	{
	    Scanner s = new Scanner(System.in);
	    node head = takeInput(s);
	    head = reverse(head);
	    System.out.println("revesed linked list: ");
	    printLL(head);
	}
	
	public static node reverse(node head) {
		node prev = null, next = null;
		node curr = head;
		while(curr != null) {
			next = curr.next;
			curr.next = prev;
			prev = curr;
			curr = next;
		}
		return prev;
	}

}

File: Data_Structures/LinkedList/Reverse_List/reverse_linkedlist.c
// Iterative C program to reverse a linked list 
#include <stdio.h> 
#include <stdlib.h> 

/* Link list node */
struct Node { 
	int data; 
	struct Node* next; 
}; 

/* Function to reverse the linked list */
static void reverse(struct Node** head_ref) 
{ 
	struct Node* prev = NULL; 
	struct Node* current = *head_ref; 
	struct Node* next = NULL; 
	while (current != NULL) { 
		// Store next 
		next = current->next; 

		// Reverse current node's pointer 
		current->next = prev; 

		// Move pointers one position ahead. 
		prev = current; 
		current = next; 
	} 
	*head_ref = prev; 
} 

/* Function to push a node */
void push(struct Node** head_ref, int new_data) 
{ 
	struct Node* new_node = (struct Node*)malloc(sizeof(struct Node)); 
	new_node->data = new_data; 
	new_node->next = (*head_ref); 
	(*head_ref) = new_node; 
} 

/* Function to print linked list */
void printList(struct Node* head) 
{ 
	struct Node* temp = head; 
	while (temp != NULL) { 
		printf("%d ", temp->data); 
		temp = temp->next; 
	} 
} 

/* Driver program to test above function*/
int main() 
{ 
	/* Start with the empty list */
	struct Node* head = NULL; 

    /* Getting user input for linkedlist */
    printf("Enter the size of linked list you wish to reverse: ");
    int sizeof_linkedlist ;
    scanf("%d",&sizeof_linkedlist);

    printf("Enter space separated inputs for the linked list you wish to reverse: ");
    int input;
    for(int i = 0 ; i<sizeof_linkedlist; i++){
        scanf("%d",&input);
        push(&head,input);
    }

	printf("Given linked list\n"); 
	printList(head); 
	reverse(&head); 
	printf("\nReversed Linked list \n"); 
	printList(head); 
	getchar(); 
} 
